File: tools/ghidra/scripts/extract_cloud_handlers.py
# Ghidra headless script: Decompile cloud/MQTT property handlers
# Find how thing.service.property.set commands flow to ISP control
# Also finds XUID internal command dispatch

import os
import sys
from ghidra.app.decompiler import DecompInterface
from ghidra.util.task import ConsoleTaskMonitor

args = getScriptArgs()
output_dir = args[0] if args else "."

program = getCurrentProgram()
fm = program.getFunctionManager()
monitor = ConsoleTaskMonitor()

decomp = DecompInterface()
decomp.openProgram(program)

print("[*] Cloud handler analysis starting...")

# Patterns to match
cloud_patterns = [
    "property_set",
    "property_get",
    "thing_service",
    "alink_property",
    "linkkit_",
    "ali_iot",
    "ali_sdk",
    "ali_link",
    "mqtt_",
    "iot_",
    "cloud_",
    "danale_",
    "tutk_",
]

xuid_patterns = [
    "xuid",
    "hi_xuid",
]

control_patterns = [
    "nightvision",
    "night_vision",
    "flip_mirror",
    "set_flip",
    "set_mirror",
    "set_led",
    "alarm_attr",
    "detect_attr",
    "detect_mode",
    "stream_attr",
    "switch_resolution",
    "set_idr",
    "draw_osd",
    "reboot",
    "watchdog",
    "factory_reset",
    "upgrade",
]

found_funcs = {}

for func in fm.getFunctions(True):
    name = func.getName()
    name_lower = name.lower()
    
    matched = False
    for p in cloud_patterns + xuid_patterns + control_patterns:
        if p in name_lower:
            matched = True
            break
    
    if matched:
        found_funcs[name] = func

print("[*] Found %d cloud/control functions to decompile" % len(found_funcs))

results = []
property_handlers = []

for name in sorted(found_funcs.keys()):
    func = found_funcs[name]
    addr = func.getEntryPoint()
    size = func.getBody().getNumAddresses()
    
    results.append("=" * 70)
    results.append("FUNCTION: %s" % name)
    results.append("ADDRESS:  %s" % addr)
    results.append("SIZE:     %d bytes" % size)
    results.append("=" * 70)
    
    try:
        res = decomp.decompileFunction(func, 60, monitor)
        if res and res.getDecompiledFunction():
            c_code = res.getDecompiledFunction().getC()
            results.append(c_code)
            
            # Look for property name strings and dispatch patterns
            for line in c_code.split("\n"):
                ls = line.strip()
                if any(kw in ls for kw in ['"NightVision', '"ImageFlip', '"MotionDetect',
                    '"AlarmSwitch', '"Floodlight', '"StatusLight', '"StorageRecord',
                    '"StreamVideo', '"SubStream', '"IRLight', '"WhiteLight',
                    '"FaceDetect', '"CrossLine', '"RegionDetect', '"IvpAbility',
                    '"IntelligentTrack', '"CustomCmd', '"RebootSchedule',
                    'property', 'XUID']):
                    property_handlers.append("%s: %s" % (name, ls))
        else:
            results.append("DECOMPILATION FAILED")
    except Exception as e:
        results.append("ERROR: %s" % str(e))
    
    results.append("")

outpath = os.path.join(output_dir, "cloud_handler_analysis.txt")
with open(outpath, "w") as f:
    f.write("CLOUD / MQTT / XUID HANDLER ANALYSIS\n")
    f.write("Binary: superb\n")
    f.write("Functions decompiled: %d\n" % len(found_funcs))
    f.write("=" * 70 + "\n\n")
    for line in results:
        f.write(line + "\n")

proppath = os.path.join(output_dir, "property_dispatch.txt")
with open(proppath, "w") as f:
    f.write("PROPERTY/XUID DISPATCH REFERENCES\n")
    f.write("=" * 70 + "\n\n")
    for line in property_handlers:
        f.write(line + "\n")

print("[*] Wrote cloud_handler_analysis.txt (%d functions)" % len(found_funcs))
print("[*] Wrote property_dispatch.txt (%d entries)" % len(property_handlers))

decomp.dispose()
print("[*] extract_cloud_handlers.py complete")

File: tools/ghidra/scripts/ExtractCloudHandlers.java
// Ghidra headless script: Decompile cloud/MQTT/XUID handler functions
//@category Analysis

import ghidra.app.script.GhidraScript;
import ghidra.app.decompiler.*;
import ghidra.program.model.listing.*;
import java.io.*;
import java.util.*;

public class ExtractCloudHandlers extends GhidraScript {

    @Override
    public void run() throws Exception {
        String[] args = getScriptArgs();
        String outputDir = args.length > 0 ? args[0] : ".";

        FunctionManager fm = currentProgram.getFunctionManager();
        DecompInterface decomp = new DecompInterface();
        decomp.openProgram(currentProgram);

        String[] patterns = {
            "property_set", "property_get", "thing_service",
            "alink_", "linkkit_", "ali_iot", "ali_sdk", "ali_link",
            "mqtt_", "iot_dm_", "iot_gateway",
            "cloud_", "danale_", "tutk_",
            "xuid", "hi_xuid",
            "nightvision", "night_vision",
            "flip_mirror", "set_flip", "set_mirror",
            "set_led", "alarm_attr", "detect_attr", "detect_mode",
            "stream_attr", "switch_resolution", "set_idr",
            "draw_osd", "factory_reset",
            "slave_reboot", "keep_alive",
        };

        Map<String, Function> targets = new TreeMap<>();

        FunctionIterator funcIter = fm.getFunctions(true);
        while (funcIter.hasNext() && !monitor.isCancelled()) {
            Function func = funcIter.next();
            String name = func.getName();
            String nameLower = name.toLowerCase();

            for (String p : patterns) {
                if (nameLower.contains(p)) {
                    targets.put(name, func);
                    break;
                }
            }
        }

        println("[*] Found " + targets.size() + " cloud/control functions to decompile");

        StringBuilder sb = new StringBuilder();
        List<String> propertyRefs = new ArrayList<>();

        sb.append("CLOUD / MQTT / XUID HANDLER ANALYSIS\n");
        sb.append("Binary: superb\n");
        sb.append("Functions decompiled: ").append(targets.size()).append("\n");
        sb.append("=".repeat(70)).append("\n\n");

        for (Map.Entry<String, Function> entry : targets.entrySet()) {
            String name = entry.getKey();
            Function func = entry.getValue();

            sb.append("=".repeat(70)).append("\n");
            sb.append("FUNCTION: ").append(name).append("\n");
            sb.append("ADDRESS:  ").append(func.getEntryPoint()).append("\n");
            sb.append("SIZE:     ").append(func.getBody().getNumAddresses()).append(" bytes\n");
            sb.append("=".repeat(70)).append("\n");

            DecompileResults res = decomp.decompileFunction(func, 60, monitor);
            if (res != null && res.decompileCompleted() && res.getDecompiledFunction() != null) {
                String cCode = res.getDecompiledFunction().getC();
                sb.append(cCode).append("\n");

                // Look for property name references
                String[] propKeywords = {
                    "NightVision", "ImageFlip", "MotionDetect", "AlarmSwitch",
                    "Floodlight", "StatusLight", "StorageRecord", "StreamVideo",
                    "SubStream", "IRLight", "WhiteLight", "FaceDetect",
                    "CrossLine", "RegionDetect", "IvpAbility", "IntelligentTrack",
                    "CustomCmd", "RebootSchedule", "XUID", "property"
                };
                for (String line : cCode.split("\n")) {
                    String trimmed = line.trim();
                    for (String kw : propKeywords) {
                        if (trimmed.contains(kw)) {
                            propertyRefs.add(name + ": " + trimmed);
                            break;
                        }
                    }
                }
            } else {
                sb.append("DECOMPILATION FAILED\n");
            }
            sb.append("\n");
        }

        PrintWriter pw = new PrintWriter(new FileWriter(new File(outputDir, "cloud_handler_analysis.txt")));
        pw.print(sb.toString());
        pw.close();

        pw = new PrintWriter(new FileWriter(new File(outputDir, "property_dispatch.txt")));
        pw.println("PROPERTY/XUID DISPATCH REFERENCES");
        pw.println("=".repeat(70));
        pw.println();
        for (String line : propertyRefs) {
            pw.println(line);
        }
        pw.close();

        println("[*] Wrote cloud_handler_analysis.txt (" + targets.size() + " functions)");
        println("[*] Wrote property_dispatch.txt (" + propertyRefs.size() + " entries)");

        decomp.dispose();
        println("[*] ExtractCloudHandlers complete");
    }
}
